prov/efa/src/efa_av.c

```diff
@@ -1075,3 +1075,97 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
 	free(av);
 	return ret;
 }
+
+#ifdef EFA_UNIT_TEST
+
+#include "efa_unit_tests.h"
+
+/* Common Setup/Teardown */
+
+struct efa_resource {
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_av *av;
+};
+
+struct efa_resource *create_efa_resource() {
+    struct efa_resource *resource = (struct efa_resource*) malloc(sizeof(struct efa_resource));
+    resource->fabric = NULL;
+    resource->domain = NULL;
+    resource->av = NULL;
+
+    struct fi_info *info;
+    int rv = fi_getinfo(FI_VERSION(1, 14), NULL, NULL, 0ULL, NULL, &info);
+    if (rv) {
+        free(resource);
+        return NULL;
+    }
+
+    rv = fi_fabric(info->fabric_attr, &resource->fabric, NULL);
+    if (rv) {
+        free(resource);
+        return NULL;
+    }
+
+    rv = fi_domain(resource->fabric, info, &resource->domain, NULL);
+    if (rv) {
+        fi_close((fid_t) resource->fabric);
+        free(resource);
+        return NULL;
+    }
+
+    struct fi_av_attr av_attr = {0};
+    rv = fi_av_open(resource->domain, &av_attr, &resource->av, NULL);
+    if (rv) {
+        fi_close((fid_t) resource->domain);
+        fi_close((fid_t) resource->fabric);
+        free(resource);
+        return NULL;
+    }
+
+    return resource;
+}
+
+void delete_efa_resource(struct efa_resource *resource) {
+    fi_close((fid_t) resource->av);
+    fi_close((fid_t) resource->domain);
+    fi_close((fid_t) resource->fabric);
+    free(resource);
+}
+
+/*
+ * Only works on nodes with EFA devices
+ * This test calls efa_ah_alloc twice with the same GID,
+ * and verifies that ibv_create_ah only gets called once.
+ */
+void create_efa_ah_twice() {
+    /* Setup Local Variables */
+    struct efa_resource *resource = create_efa_resource();
+    assert_non_null(resource);
+    struct efa_av *av = container_of(resource->av, struct efa_av, util_av.av_fid);
+    assert_non_null(av);
+    uint8_t gid[EFA_GID_LEN];
+    memset(gid, 7, EFA_GID_LEN);
+
+    /* Set mock expectations and call UUT, success */
+    int ibv_rv = 4242;
+    expect_value(__wrap_ibv_create_ah, pd, av->domain->ibv_pd);
+    expect_any(__wrap_ibv_create_ah, attr);
+    expect_value(__wrap_efadv_query_ah, ibvah, ibv_rv);
+    expect_any(__wrap_efadv_query_ah, attr);
+    expect_any(__wrap_efadv_query_ah, inlen);
+    will_return(__wrap_ibv_create_ah, ibv_rv);
+    will_return(__wrap_efadv_query_ah, 0);
+    struct efa_ah *efa_ah = efa_ah_alloc(av, gid);
+    assert_non_null(efa_ah);
+
+    /* Call UUT again */
+    struct efa_ah *efa_ah2 = efa_ah_alloc(av, gid);
+    assert_true(efa_ah2 == efa_ah);
+    assert_true(efa_ah->refcnt == 2);
+    assert_true(av->ah_map->refcnt == 2);
+
+    delete_efa_resource(resource);
+    av = NULL;
```

Comment:
This seems pointless as we're exiting the context of the function anyway

Reply:
done